rig/machine_control/machine_controller.py

```diff
@@ -1657,51 +1659,120 @@ def get_p2p_routing_table(self, x, y):
+        if the remote device is not a SpiNNaker chip, the link is reported as
+        dead.
+
+        See also: :py:meth:`.get_chip_info`.
 
         Returns
         -------
         set([:py:class:`rig.machine.Links`, ...])
         """
-        link_up = self.read_struct_field("sv", "link_up", x, y)
-        return set(link for link in Links if link_up & (1 << link))
+        return self.get_chip_info(x, y).working_links
 
     @ContextMixin.use_contextual_arguments()
     def get_num_working_cores(self, x, y):
-        """Return the number of working cores, including the monitor."""
+        """Return the number of working cores, including the monitor.
+
+        See also: :py:meth:`.get_chip_info`.
+        """
         return self.read_struct_field("sv", "num_cpus", x, y)
 
-    def get_machine(self, x=0, y=0, default_num_cores=18):
+    def _get_all_chips_info(self, x=0, y=0):
+        """For internal use only. Get system dimensions and a ChipInfo for all
+        working chips.
+
+        Returns
+        -------
+        (width, height, chips_info)
+            width and height are integers giving the dimensions of the system.
+
+            chips_info is a dictionary {(x, y): :py:class:`.ChipInfo`, ...}
+            giving the chip info for all chips which are reported as working in
+            the P2P table and which responded to :py:meth:`.get_chip_info`.
+        """
+        # A quick way of getting a list of working chips
+        p2p_tables = self.get_p2p_routing_table(x, y)
+
+        # Calculate the extent of the system
+        max_x = max(x_ for (x_, y_), r in iteritems(p2p_tables)
+                    if r != consts.P2PTableEntry.none)
+        max_y = max(y_ for (x_, y_), r in iteritems(p2p_tables)
+                    if r != consts.P2PTableEntry.none)
+
+        chips_info = {}
+
+        for (x, y), p2p_route in iteritems(p2p_tables):
+            if x <= max_x and y <= max_y:
+                if p2p_route != consts.P2PTableEntry.none:
+                    try:
+                        chips_info[(x, y)] = self.get_chip_info(x, y)
+                    except SCPError:
+                        # The chip was listed in the P2P table but is not
+                        # responding. Assume it is dead and don't include it in
+                        # the info returned.
+                        pass
+
+        return (max_x + 1, max_y + 1, chips_info)
+
+    def get_machine(self, x=0, y=0, default_num_cores=18,
+                    _all_chips_info=None):
         """Probe the machine to discover which cores and links are working.
 
         .. note::
             Links are reported as dead when the device at the other end of the
-            link does not respond to SpiNNaker nearest neighbour packets. This
-            may thus mistakenly report links attached to peripherals as dead.
+            link is not a SpiNNaker chip. Thus links attached to peripherals
+            are always marked as dead.
```

Comment:
I feel this needs clarifying to the extent that a link is reported as dead *when it is dead* or when the device at the other end is not a SpiNNaker chip. But I could just be being really picky, it is my way.